path_collector.go

```diff
@@ -230,3 +241,96 @@ func ExcludeWithGitignore(gitignorePath string, paths []string) ([]string, error
 	logger.Debug(logger.DebugCodePaths, "paths to format: %s", pathsToFormat)
 	return pathsToFormat, nil
 }
+
+const DefaultPatternFile = "yamlfmt.patterns"
+
+// PatternFile determines which files to format and which to ignore based on a pattern file in gitignore(5) syntax.
+type PatternFile struct {
+	fs      fs.FS
+	matcher *ignore.GitIgnore
+}
+
+// NewPatternFile initializes a new PatternFile using the provided file(s).
+// If multiple files are provided, their content is concatenated in order.
+// All patterns are relative to the current working directory.
+func NewPatternFile(files ...string) (PatternFile, error) {
+	r, err := cat(files...)
+	if err != nil {
+		return PatternFile{}, err
+	}
+
+	wd, err := os.Getwd()
+	if err != nil {
+		return PatternFile{}, fmt.Errorf("os.Getwd: %w", err)
+	}
+
+	return NewPatternFileFS(r, os.DirFS(wd)), nil
+}
+
+// cat concatenates the contents of all files in its argument list.
+func cat(files ...string) (io.Reader, error) {
+	var b bytes.Buffer
+
+	for _, f := range files {
+		fh, err := os.Open(f)
+		if err != nil {
+			return nil, err
+		}
+		defer fh.Close()
+
+		if _, err := io.Copy(&b, fh); err != nil {
+			return nil, fmt.Errorf("copying %q: %w", f, err)
+		}
+		fh.Close()
+
+		// Append a newline to avoid issues with files lacking a newline at end-of-file.
+		fmt.Fprintln(&b)
+	}
+
+	return &b, nil
+}
+
+// NewPatternFileFS reads a pattern file from r and uses fs for file lookups.
+// It is used by NewPatternFile and primarily public because it is useful for testing.
+func NewPatternFileFS(r io.Reader, fs fs.FS) PatternFile {
+	var lines []string
+
+	s := bufio.NewScanner(r)
+	for s.Scan() {
+		lines = append(lines, s.Text())
+	}
+
+	return PatternFile{
+		fs:      fs,
+		matcher: ignore.CompileIgnoreLines(lines...),
+	}
+}
+
+// CollectPaths implements the PathCollector interface.
+func (pf PatternFile) CollectPaths() ([]string, error) {
```

Comment:
Is it fine to implement this on `*PatternFile` rather than `PatternFile` to match the other `PathCollector` implementations?

Reply:
✅ done